xds/src/test/java/io/grpc/xds/XdsClientWrapperForServerSdsTest.java

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.mockito.internal.verification.VerificationModeFactory.times;
+
+import com.google.common.base.Strings;
+import io.envoyproxy.envoy.api.v2.auth.DownstreamTlsContext;
+import io.grpc.xds.internal.sds.CommonTlsContextTestsUtil;
+import io.netty.channel.Channel;
+import java.io.IOException;
+import java.net.Inet4Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+/** Tests for {@link XdsClientWrapperForServerSds}. */
+@RunWith(JUnit4.class)
+public class XdsClientWrapperForServerSdsTest {
+
+  private static final int PORT = 7000;
+
+  @Mock private XdsClient xdsClient;
+  @Mock private Channel channel;
+
+  private XdsClientWrapperForServerSds xdsClientWrapperForServerSds;
+  private DownstreamTlsContext tlsContext1;
+  private DownstreamTlsContext tlsContext2;
+
+  @Before
+  public void setUp() throws IOException {
+    MockitoAnnotations.initMocks(this);
+    xdsClientWrapperForServerSds = new XdsClientWrapperForServerSds(PORT, xdsClient);
+    tlsContext1 = CommonTlsContextTestsUtil.buildTestDownstreamTlsContext("CERT1", "VA1");
+    tlsContext2 = CommonTlsContextTestsUtil.buildTestDownstreamTlsContext("CERT2", "VA2");
+  }
+
+  @Test
+  public void verifyListenerWatcherRegistered() {
+    verify(xdsClient, times(1)).watchListenerData(eq(PORT), any(XdsClient.ListenerWatcher.class));
+  }
+
+  @Test
+  public void listener_nullFilterChainMatch1() throws UnknownHostException {
+    // filterChainMatch missing for filter1, so it matches and returns tlsContext1
+    commonFilterChainMatchTest(-1, null, null, "192.168.10.1", "192.168.10.2", tlsContext1);
```

Comment:
this is harder to read. consider parameterized test.

Reply:
You mean `org.junit.runners.Parameterized` ? I prefer adding more comments to the existing code to make it more readable.

Reply:
requires comment means it is not readable. the code structure is already ready to be converted to parameterized. anyways, i'll leave it up to you as long as it is readable.

Reply:
With parameterized, it is helpful if parameter values tell you what the test is. In this case one needs to know more. e.g. passing null for certain params doesn't convey anything. But I'll give it some more thought.